lua/plenary/scandir.lua

```diff
@@ -175,57 +175,72 @@ m.scan_dir_async = function(path, opts)
+
+    while nr ~= 0 do
+      octal = octal + (nr % 8) * i
+      nr = math.floor(nr / 8)
+      i = i * 10
+    end
+
+    return octal;
   end
 
-  return octal;
-end
+  local type_tbl = { [1]  = 'p', [2]  = 'c', [4]  = 'd', [6]  = 'b', [10] = '.', [12] = 'l', [14] = 's' }
+  local permissions_tbl = { [0] = '---', '--x', '-w-', '-wx', 'r--', 'r-x', 'rw-', 'rwx' }
+  local bit_tbl = { 4, 2, 1 }
 
-local type_tbl = { [1]  = 'p', [2]  = 'c', [4]  = 'd', [6]  = 'b', [10] = '-', [12] = 'l', [14] = 's' }
-local permissions_tbl = { [0] = '---', '--x', '-w-', '-wx', 'r--', 'r-x', 'rw-', 'rwx' }
-local bit_tbl = { 4, 2, 1 }
-
-local gen_permissions = function(stat)
-  local octal = string.format('%6d', conv_to_octal(stat.mode))
-  local l4 = octal:sub(#octal - 3, -1)
-  local bit = tonumber(l4:sub(1, 1))
-
-  local result = type_tbl[tonumber(octal:sub(1, 2))] or '-'
-  for i = 2, #l4 do
-    result = result .. permissions_tbl[tonumber(l4:sub(i, i))]
-    if bit - bit_tbl[i - 1] >= 0 then
-      result = result:sub(1, -2) .. (bit_tbl[i - 1] == 1 and 't' or 's')
-      bit = bit - bit_tbl[i - 1]
+  return function(cache, mode)
+    if cache[mode] then return cache[mode] end
+
+    local octal = string.format('%6d', conv_to_octal(mode))
+    local l4 = octal:sub(#octal - 3, -1)
+    local bit = tonumber(l4:sub(1, 1))
+
+    local result = type_tbl[tonumber(octal:sub(1, 2))] or '-'
+    for i = 2, #l4 do
+      result = result .. permissions_tbl[tonumber(l4:sub(i, i))]
+      if bit - bit_tbl[i - 1] >= 0 then
+        result = result:sub(1, -2) .. (bit_tbl[i - 1] == 1 and 'T' or 'S')
+        bit = bit - bit_tbl[i - 1]
+      end
     end
+
+    cache[mode] = result
+    return result
   end
-  return result
-end
+end)()
 
-local gen_size = function(stat)
-  local size = stat.size
-  for _, v in ipairs{ '', 'K', 'M', 'G', 'T', 'P', 'E', 'Z' } do
-    if math.abs(size) < 1024.0 then
-      if math.abs(size) > 9 then
-        return string.format("%3d%s", size, v)
-      else
-        return string.format("%3.1f%s", size, v)
+local gen_size = (function()
+  local size_types = { '', 'K', 'M', 'G', 'T', 'P', 'E', 'Z' }
+
+  return function(size)
+    -- TODO(conni2461): If type directory we could just return 4.0K
+    for _, v in ipairs(size_types) do
+      if math.abs(size) < 1024.0 then
+        if math.abs(size) > 9 then
+          return string.format("%3d%s", size, v)
+        else
+          return string.format("%3.1f%s", size, v)
+        end
       end
+      size = size / 1024.0
     end
-    size = size / 1024.0
+    return string.format("%.1f%s", size, 'Y')
   end
-  return string.format("%.1f%s", size, 'Y')
-end
+end)()
 
-local gen_date = function(stat)
-  return os.date('%b %d %H:%M', stat.mtime.sec)
-end
+local gen_date = (function()
+  local current_year = os.date('%Y')
+  return function(mtime)
+    if current_year ~= os.date('%Y', mtime) then
+      return os.date('%b %d  %Y', mtime)
+    end
+    return os.date('%b %d %H:%M', mtime)
+  end
```

Comment:
```suggestion
    local invalid  = current_year ~= os.date('%Y', mtime)
    return invalid and os.date('%b %d  %Y', mtime) or os.date('%b %d %H:%M', mtime)
  end
```

:D I know you won't lie this but it's better then if then.

Reply:
I think i keep the original :)

Reply:
I know it :P